content/ros/ROS_Nodes_Topics_Messages/README.md

```diff
@@ -0,0 +1,280 @@
+```
+rosrun my_first_package 01_simple_node.py
+```
+
+
+<details><summary>More details:</summary>
+
+<br/>
+
+- ```rosrun``` searches withing the directory tree of the given package. The scripts doesn't even have to be located in ```scripts``` folder. It's only common between ROS users to place Python scripts inside the ```scripts``` folder.
+  
+- :warning:  ```rosrun``` will not find your Python scripts if you don't change their access permession to executable.
+
+- ```rosrun``` takes options and more arguments that will be covered later.
+
+### How does ```rosrun``` find my files?
+It searches the directories listed in the ```ROS_PACKAGE_PATH``` environment variable. Try to see to where it points at:
+
+```
+echo $ROS_PACKAGE_PATH
+```
+
+### ```rosnode``` command:
+- List all the running nodes:
+```
+rosnode list
+```
+- Get more info. about a certain node:
+```
+rosnode info <node name>
+```
+- Others:
+```
+rosnode kill <node name>
+rosnode ping <node name>
+```
+...etc
+</details>
+
+
+## Some of ROS commands:
+- put the script (```02_simple_publisher.py```) inside our package (```my_first_package/scripts```). Give it execution permission.
+
+- Run the script using ```rosrun```.
+```
+rosrun my_first_package 02_simple_publisher.py
+```
+
+- Let's see the current list of topics.  To do that we will use ```rostopic list``` command:
+```
+rostopic list
+```
+- Let's see the messages we are publishing. To do that we will use ```rostopic echo``` command:
+
+```
+rostopic echo /myFirstTopic
+```
+
+<details><summary>More details:</summary>
+
+- Get field type of a topic:
+```
+rostopic type <topic name>
+```
+- Get information about ROS a topic:
+```
+rostopic info <topic name>
+```
+- Display publishing rate of topic:
+```
+rostopic hz <topic name>
+```
+<br/>
+</details>
+
+
+## Simple subscriber
+
+- Run our subscriber ```04_simple_subscriber.py``` node (after doing necessary steps):
+  
+```
+rosrun my_first_package 04_simple_subscriber.py
+```
+- The script is listening but there are no messages being published.
+- Run our publisher node in a second terminal:
+```
+rosrun my_first_package 02_simple_publisher.py
+```
+- You can also publish from the terminal directly using ```rostopic pub``` command:
+```
+rostopic pub /myFirstTopic std_msgs/String "data: 'hey there'"
+```
+
+
+
+## Using launch files
+
+- Create a ```launch``` folder in your package ```my_first_package```.
+- Place our launch files in it.
+- Use ```roslaunch``` command to run ```01_simple.launch```:
+
+```
+roslaunch my_first_package 01_simple.launch
+```
+
+- Check topic names, and notice how our node name was changed.
+```
+rosnode list
+```
+<details><summary>More details:</summary>
+
+- Launch files don't necessarily have to be placed in a ```launch``` folder, they just have to be placed anywhere inside the package. ```roslaunch``` command will find them.
+
+- When you run a launch file, it also brings up the master, so you don't have to run ```roscore``` command
```

Comment:
With the caveat that, if the roscore is not run separately, the ROS master will be dead when the launcher is stopped.

Reply:
done